docs/templates/block-template-controller.md

```diff
@@ -0,0 +1,77 @@
+* [render_block_template()](#render_block_template)
+
+## Overview
+
+In the initialization of the class, we hook into the three hooks listed below. These provide us with all of the extensibility points necessary in order to load our own block templates alongside the themes.
+
+Within each method section, I will explain in what scenarios they are executed. 
+
+* filter: `get_block_templates` with `add_block_templates`.
+* filter: `pre_get_block_file_template` with `get_block_file_template`.
+* action: `template_redirect` with `render_block_template`.
+
+## add_block_templates( $query_result, $query, $template_type )
+
+This method is applied to the filter `get_block_templates`, which is executed before returning a unified list of template objects based on a query.
+
+**Typically executed when:**
+* Loading the "All Templates" view in the Site Editor
+* Loading one of the templates on the frontend where the query would build a list of relevant templates based on a hierarchy (for example, the product page hierarchy could be an array containing  `single-product-[product-name].html`, `single-product.html`, `single.html`).
+
+**This method is responsible for:**
+
+* Giving our templates a user-friendly title (e.g. turning "single-product" into "Product Page").
+* It collects all the WooCommerce templates from both the filesystem and the database (customized templates are stored in the database as posts) and adds them to the returned list.
+* In the event the theme has a `archive-product.html` template file, but not category/tag template files, it is eligible to use the `archive-product.html` file in their place. So we trick Gutenberg in thinking some templates (e.g. category/tag) have a theme file available if it is using the `archive-product.html` template, even though _technically_ the theme does not have a specific file for them.
+* If the query is targeting templates for a specific post type (e.g. on the Edit Post view), we do not add ineligible WooCommerce templates to the returned list.
+
+### Return value
+
+This method will return an array of `WP_Block_Template` values
+
+## get_block_file_template( $template, $id, $template_type )
+
+This method is applied to the filter `pre_get_block_file_template` inside the WordPress core function `get_block_file_template` (not to be confused with this method from the BlockTemplateController, of which has the same name). 
+
+The order of execution is as follows:
+
+1. `get_block_template()` from WordPress core will execute, and attempt to retrieve a customized version of the template from the database.
+2. If it fails to retrieve one, it will execute the `get_block_file_template()` from WordPress core which will apply the filter `pre_get_block_file_template` which is what we hook into to return our template file, which will trigger an early return to prevent WordPress from continuing its query.
```

Comment:
Any way we can contain the “whiches”? So many relative clauses are a bit hard to parse, especially as a non native speaker, I suppose.

One way I can think of would be to just stop the sentence and resume another one, something like:

_If it fails to retrieve one, it will execute the `get_block_file_template()` from WordPress core which will apply the filter `pre_get_block_file_template`. This is were we hook into to return our template file, and trigger an early return to prevent WordPress from continuing its query_

Reply:
Nice! Reads much better